storage: add new CheckSSTConflicts randomized test

jbowens opened the pull request:
**enginepb: add MVCCStats.Formatted**

Move the storage test formatStats function onto the MVCCStats type itself in
preparation for using it in additional places.

**storage: change CheckSSTConflicts start and end types**

Callers of CheckSSTConflicts pass in start and end boundaries for the sstable.
Previously, these were passed as MVCCKeys, although the end key's timestamp was
ignored. This resulted in confusing semantics whereby the end boundary was
interpreted as exclusive end bound of `end.Key`.

**storage: add new CheckSSTConflicts randomized test**

Add a new randomized test exercising CheckSSTConflicts. Additionally, sketch
out a `meta` package to aid in writing randomized tests like this one. In the
future, this package may be extracted to the cockroachdb/metamorphic
repository.

Epic: None
Release note: None
Informs #94141.
Informs cockroachdb/pebble#2086.

jbowens committed:
storage: add new CheckSSTConflicts randomized test

Add a new randomized test exercising CheckSSTConflicts. Additionally, sketch
out a `meta` package to aid in writing randomized tests like this one. In the
future, this package may be extracted to the cockroachdb/metamorphic
repository.

Epic: None
Release note: None

jbowens committed:
enginepb: add MVCCStats.Formatted

Move the storage test formatStats function onto the MVCCStats type itself in
preparation for using it in additional places.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98408)
<!-- Reviewable:end -->

jbowens reviewed (commented):
This test is currently failing. I'm hoping it's failing due to #94141 and the test can be used to produce a simpler reproduction of that issue.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98408#-:-NQCLbDD1dWPwrUX3ozK:b-j9b5ng)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98408)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @sumeerbhola)

___
*[`pkg/storage/meta/run.go` line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98408#-NQLt067DRhanbGXs4BS:-NQLt067DRhanbGXs4BT:b-72vi25) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b1fec5f6ce6e0bc9f7b80ad6bcb661c2b789912/pkg/storage/meta/run.go#L58)):*
> ```Go
> // each item. It's intended to be used with a function returned by Weighted,
> // whhere the item itself is func(rng *rand.Rand).
> func Generate[I any](rng *rand.Rand, n int, fn func(*rand.Rand) func(*rand.Rand) I) []I {
> ```

I don't get the generic-ness of `ItemWeight` if we're only going to use them with this function which assumes that the items are a `func(*rand.Rand) I`.

Can't we have a `GenWithWeight[I] struct { Gen func(*rand.Rand) I; Weight int }` and pass a `...GenWithWeight[I]` to `Generate` ? Are we envisioning any other kind of usage?


<!-- Sent from Reviewable.io -->

jbowens reviewed (commented):
@itsbilal — looks like this is still failing after rebasing over master, including #98426.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98408#-:-NQQXoDH7vUwwkc0lFQn:b-4cfnja)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @sumeerbhola)

___
*[`pkg/storage/meta/run.go` line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98408#-NQLt067DRhanbGXs4BS:-NQQX5slA0G01drDZ7Ql:bdtae64) ([raw file](https://github.com/cockroachdb/cockroach/blob/d15ea19a9449e80c14f136d3fb9228b178a88c73/pkg/storage/meta/run.go#L58)):*
>  Are we envisioning any other kind of usage?

Yes. Picking among weighted items is expected to be useful during generation of the individual options themselves as well. For example, the Pebble metamorphic tests [pick among IterOptions.KeyTypes enum values](https://github.com/cockroachdb/pebble/blob/51422ae2d449f51cbbd79e36238b3147dce257c3/internal/metamorphic/generator.go#L566-L577).


<!-- Sent from Reviewable.io -->

jbowens reviewed (commented):
Looks like this one is still failing, even if kvnemesis isn't :thinking: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98408#-:-NQWEjg01sPreR-h5KfD:b-ntujl3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @sumeerbhola)



<!-- Sent from Reviewable.io -->

itsbilal commented:
@jbowens I realized that the randomized test produces sst range keys that are of different timestamps. While the function supports that, it's unlikely (impossible I believe?) for that to happen with practical uses of `AddSSTable`; the entirety of the SST should be of the same timestamp. Maybe someone in KV can confirm, but at least this sort of a failure doesn't point to something imminently serious.

jbowens commented:
I'm forgetting, do we support RESTORE-ing with history, or only backing up with history and restoring the most recent values? If we RESTORE with history, that must use AddSSTable without setting the `sstTimestamp` parameter.

I suppose the disaggregated storage RESTORE will need to restore with history or perform read-time timestamp rewriting, but that'll be for 23.2 in any case.

jbowens commented:
@itsbilal I asked in #disaster-recovery and it sounds like both cluster-to-cluster streaming and tenant restores both use AddSSTable with varied timestamps.

itsbilal reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo some comments, thanks so much for doing this!

> Reviewed 1 of 2 files at r1, 7 of 7 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98408#-:-NRiD4EF77KZ2IQLT0_1:bwpjupi)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jbowens, @RaduBerinde, and @sumeerbhola)

___
*[`pkg/storage/meta/run.go` line 74 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/98408#-NRiC0HO0iLW4_jCzW2e:-NRiC0HO0iLW4_jCzW2f:b-b2emi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad3ddc7372b7818335b38c4dd7eb3b3a6c765ede/pkg/storage/meta/run.go#L74)):*
> ```Go
> 	l.wIndent = newlineIndentingWriter{Writer: l.w, indent: []byte("  ")}
> 
> 	// Ensure pancis result in printing the history.
> ```

nit: panics
___
*[`pkg/storage/sst_test.go` line 415 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/98408#-NRiC_viDW3Gchwpw9px:-NRiC_viDW3Gchwpw9py:b-9h320t) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad3ddc7372b7818335b38c4dd7eb3b3a6c765ede/pkg/storage/sst_test.go#L415)):*
> ```Go
> 			rk.EndKey, rk.StartKey = rk.StartKey, rk.EndKey
> 		} else if v == 0 {
> 			rk.StartKey = append(append(rk.EndKey[:0], rk.StartKey...), 0x00)
> ```

Shouldn't this be `rk.EndKey = ...` ?


<!-- Sent from Reviewable.io -->

jbowens reviewed (commented):
Good call. The new CI job picked up a new failure :S

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98408#-:-NRiwRfuEZYqbEnh2YWX:bcxc9gz)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @RaduBerinde, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

itsbilal commented:
In the process of stressing this test for `100347`, I'd sometimes see this error in this test:

```
=== RUN   TestCheckSSTConflictsRandomized
    test_log_scope.go:161: test logs captured to: /Users/bilal/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/d1091de82354dd1e885605d51bd315be/logTestCheckSSTConflictsRandomized3309570774
    test_log_scope.go:79: use -show-logs to present logs inline
    sst_test.go:184: Using seed 8730232687463734057; to re-run set COCKROACH_RANDOM_SEED=8730232687463734057
    run.go:148: iterator with constraint=2 is being used with key "\x00" that has constraint=1
    run.go:155: History:

        op      0: MVCCDeleteRangeUsingTombstone("\x00", "nfxem", 0.000000105,0) = ok
        op      1: MVCCPut("nvt"@107, "hb") = ok
        op      2: MVCCPut("nebi"@174, "s") = ok
        op      3: MVCCPut("nlsj"@156, "wrbiqenq") = ok
        op      4: MVCCDelete("nzou"@82) = ok
        op      5: MVCCPut("nsxu"@91, "gsjaxoq") = ok
        op      6: MVCCDelete("nagjk"@168) = ok
        op      7: MVCCPut("nnasi"@103, "nbkdncf") = ok
        op      8: MVCCPut("nv"@136, "") = ok
        op      9: CheckSSTConflicts(disallowShadowing=0.000000161,0, usePrefix=true) = error: iterator with constraint=2 is being used with key "\x00" that has constraint=1
```

It looks like some of the random keys we generate are straddling the local/global keyspace and are throwing off the intentInterleavingIter.

jbowens reviewed (commented):
> I'd sometimes see this error in this test, which would cause a panic and fail the test:

I fixed this—there was a sequence of during generation that could leave `rangeStart` unset, generating a key with a NUL key start bound.

With
```
./dev test -f TestCheckSSTConflictsRandomized pkg/storage --stress
```

it looks like we still have at least one more stats calculation issue :(

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98408#-:-NSBr1f60GQjtMvLhU_I:b51vaoh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @RaduBerinde, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

itsbilal commented:
This came up in the storage weekly, but the stats failures I'm seeing are involving the same key being present in the sst at different timestamps, of which one has a conflict but we don't error on the conflict (and I believe this function has never errored on this type of a conflict). Here's one example:

```
          SST:
            "ni"/0.000001035,0→/BYTES/jjub
            "ni"/0.000000970,0→/<empty>
          Before:
            "nd"/0.000000992,0→/BYTES/ljxrdz
            "ni"/0.000000970,0→/BYTES/xscxnv
          After:
            "nd"/0.000000992,0→/BYTES/ljxrdz
            "ni"/0.000001035,0→/BYTES/jjub
            "ni"/0.000000970,0→/<empty>
```

We should have errored on the ni@970 conflict but we didn't, instead our stats are incorrect as it just assumed ni@1035 is the only inbound key with the prefix "ni".

Asking disaster recovery on whether this is possible in the first place.

itsbilal commented:
In addition, I noticed that there's a subtle bug in how we set start and end when calling `CheckSSTConflicts` in the randomized test. All callers of CheckSSTConflicts pass meta keys i.e. timestamp 0 as start/end bounds of the sstable, except for the randomized test. This change fixes this:

```
diff --git a/pkg/storage/sst_test.go b/pkg/storage/sst_test.go
index 5400e12f114..95f4fc8e546 100644
--- a/pkg/storage/sst_test.go
+++ b/pkg/storage/sst_test.go
@@ -476,8 +476,8 @@ func (o *checkSSTConflictsOp) Run(l *meta.Logger, s *sstTestState) {
                ctx,
                f.Bytes(),
                s.engine,
-               s.bufferedOps[0].sortKey(),
-               s.bufferedOps[len(s.bufferedOps)-1].sortKey(),
+               MVCCKey{Key: s.bufferedOps[0].sortKey().Key},
+               MVCCKey{Key: s.bufferedOps[len(s.bufferedOps)-1].sortKey().Key},
                o.start,
                o.end,
                o.disallowShadowing != nil,
@@ -585,7 +585,7 @@ func (o *addSSTableOp) Run(l *meta.Logger, s *sstTestState) {
                ctx,
                f.Bytes(),
                s.engine,
-               s.bufferedOps[0].sortKey(),
+               MVCCKey{Key: s.bufferedOps[0].sortKey().Key},
                MVCCKey{Key: o.end},
                o.start,
                o.end,
```

itsbilal commented:
Took a look at all uses of `AddSSTable` that I could find, and I couldn't find a single one where we add multiple MVCC keys per roachpb.Key, and also do conflict checking. The c2c case, where we could add multiple keys, sets none of `DisallowConflicts`, `DisallowShadowing` or `DisallowShadowingBelow` so CheckSSTConflicts won't run there to begin with. It's why the assumption that this function has made for years (incl. pre-range keys) is safe to make.

Knowing this, I think we should update the randomized test to only output one TS per user key for point keys.

jbowens commented:
> In addition, I noticed that there's a subtle bug in how we set start and end when calling CheckSSTConflicts in the randomized test. All callers of CheckSSTConflicts pass meta keys i.e. timestamp 0 as start/end bounds of the sstable, except for the randomized test. This change fixes this:

There's a lot of subtlety there in the `CheckSSTConflicts` API. The `end` boundary doesn't make sense as a `MVCCKey` if we require `end.Key` to be an exclusive bound (not even `end` itself).

jbowens committed:
storage: change CheckSSTConflicts start and end types

Callers of CheckSSTConflicts pass in start and end boundaries for the sstable.
Previously, these were passed as MVCCKeys, although the end key's timestamp was
ignored. This resulted in confusing semantics whereby the end boundary was
interpreted as exclusive end bound of `end.Key`.

Epic: none
Release note: none

jbowens reviewed (commented):
```
4417 runs so far, 0 failures, over 21m55s
```

:sunglasses: 

Thanks for your help here @itsbilal!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98408#-:-NSMAqljCLk3LBsgfEtz:b-fk5ujb)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @RaduBerinde, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

jbowens commented:
TFTRs!

bors r+

yuzefovich commented:
On CI run:
```
Failed
=== RUN   TestCheckSSTConflictsRandomized
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c5eb8fc8b8e683c19f3c3e4238f64094/logTestCheckSSTConflictsRandomized3898596176
    test_log_scope.go:79: use -show-logs to present logs inline
    sst_test.go:183: Using seed -4844134980158906090; to re-run set COCKROACH_RANDOM_SEED=-4844134980158906090
    run.go:148: PutMVCC timestamp is empty
    run.go:155: History:
        
        op      0: MVCCDelete("diez"@0) = ok
        op      1: CheckSSTConflicts(disallowShadowing=<nil>, usePrefix=true) = error: PutMVCC timestamp is empty
    panic.go:522: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: /artifacts/tmp/_tmp/c5eb8fc8b8e683c19f3c3e4238f64094/logTestCheckSSTConflictsRandomized3898596176
--- FAIL: TestCheckSSTConflictsRandomized (0.02s)
```

bors r-

itsbilal reviewed (approved):
LGTM

Does the failing seed pass with this change?
